File: benchmarks/pilot/gold/fsharp/001-priority/Program.fs
open System
open System.Text.Json

[<CLIMutable>]
type Order =
    { id: string
      createdAt: DateTimeOffset
      status: string
      priority: Nullable<int> }

[<CLIMutable>]
type Request =
    { operation: string
      orders: Order array }

[<CLIMutable>]
type Response = { ids: string array }

let options =
    JsonSerializerOptions(
        PropertyNameCaseInsensitive = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    )

let isStatus expected actual =
    String.Equals(expected, actual, StringComparison.OrdinalIgnoreCase)

let priorityOf (order: Order) =
    if order.priority.HasValue then order.priority.Value else 0

let handle (request: Request) =
    match request.operation with
    | operation when isStatus "ready" operation ->
        request.orders
        |> Array.filter (fun order -> isStatus "pending" order.status)
        |> Array.sortBy (fun order -> -(priorityOf order), order.createdAt, order.id)
        |> Array.map (fun order -> order.id)
        |> fun ids -> { ids = ids }
    | operation -> invalidArg "operation" $"Unknown operation: {operation}"

let mutable running = true

while running do
    match Console.ReadLine() with
    | null -> running <- false
    | input ->
      try
        let request = JsonSerializer.Deserialize<Request>(input, options)
        match request with
        | null -> raise (InvalidOperationException("Request was null"))
        | request ->
          let response = handle request
          Console.WriteLine(JsonSerializer.Serialize(response, options))
      with ex ->
        Console.WriteLine(JsonSerializer.Serialize({| error = ex.Message |}, options))

File: benchmarks/pilot/gold/csharp/001-priority/Program.cs
using System.Text.Json;

namespace OrderFlow;

public sealed class Order
{
    public string Id { get; init; } = "";
    public DateTimeOffset CreatedAt { get; init; }
    public string Status { get; init; } = "";
    public int? Priority { get; init; }
}

public sealed class Request
{
    public string Operation { get; init; } = "";
    public Order[] Orders { get; init; } = [];
}

public sealed record Response(string[] Ids);

public static class Program
{
    private static readonly JsonSerializerOptions Options = new()
    {
        PropertyNameCaseInsensitive = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    };

    public static void Main()
    {
        string? line;
        while ((line = Console.ReadLine()) is not null)
        {
            try
            {
                var request = JsonSerializer.Deserialize<Request>(line, Options)
                    ?? throw new InvalidOperationException("Request was null");
                var response = Handle(request);
                Console.WriteLine(JsonSerializer.Serialize(response, Options));
            }
            catch (Exception ex)
            {
                Console.WriteLine(JsonSerializer.Serialize(new { error = ex.Message }, Options));
            }
        }
    }

    private static Response Handle(Request request)
    {
        if (!string.Equals(request.Operation, "ready", StringComparison.OrdinalIgnoreCase))
            throw new ArgumentException($"Unknown operation: {request.Operation}", nameof(request.Operation));

        var ids = request.Orders
            .Where(order => string.Equals(order.Status, "pending", StringComparison.OrdinalIgnoreCase))
            .OrderByDescending(order => order.Priority ?? 0)
            .ThenBy(order => order.CreatedAt)
            .ThenBy(order => order.Id, StringComparer.Ordinal)
            .Select(order => order.Id)
            .ToArray();
        return new Response(ids);
    }
}
